File: python/clc/services/caller_detector.py
from clc.enums import CallerType


class CallerDetector:
    SEO_BOT_PATTERNS = [
        "googlebot",
        "bingbot",
        "slurp",
        "duckduckbot",
        "baiduspider",
        "yandexbot",
        "facebookexternalhit",
        "twitterbot",
        "linkedinbot",
        "whatsapp",
        "slackchannel",
        "pinterestbot",
    ]

    def detect(self, user_agent: str = "", auth_token: str = "") -> int:
        if self._is_seo_bot(user_agent):
            return CallerType.BOT.value

        if self._is_authenticated(auth_token):
            return CallerType.AUTHENTICATED.value

        return CallerType.ATTACKER.value

    def get_label(self, caller_mask: int) -> str:
        if caller_mask == CallerType.BOT.value:
            return "SEO Bot"
        elif caller_mask == CallerType.AUTHENTICATED.value:
            return "Authenticated User"
        else:
            return "Unknown Caller"

    def _is_seo_bot(self, user_agent: str) -> bool:
        user_agent_lower = user_agent.lower()
        return any(pattern in user_agent_lower for pattern in self.SEO_BOT_PATTERNS)

    def _is_authenticated(self, auth_token: str) -> bool:
        if not auth_token:
            return False

        import re

        pattern = r"^(Bearer|Token)\s+\S+$"
        return bool(re.match(pattern, auth_token, re.IGNORECASE))

File: laravel/app/Services/CallerDetector.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Enums\CallerType;
use Illuminate\Http\Request;

/**
 * Caller detection service.
 *
 * Identifies the type of caller (SEO bot, authenticated user, or attacker)
 * based on User-Agent header and authorization token.
 *
 * Detection result determines which projection is rendered.
 */
final class CallerDetector
{
    /**
     * Known SEO bot user agent patterns.
     *
     * @var array<string>
     */
    private const SEO_BOT_PATTERNS = [
        'googlebot',
        'bingbot',
        'slurp',      // Yahoo
        'duckduckbot',
        'baiduspider',
        'yandexbot',
        'facebookexternalhit',
        'twitterbot',
        'linkedinbot',
        'whatsapp',
        'slackchannel',
        'pinterestbot',
    ];

    /**
     * Detect the caller type from the request.
     *
     * Detection priority:
     * 1. SEO bot (User-Agent match)
     * 2. Authenticated user (Authorization header)
     * 3. Default: Attacker/Unknown
     *
     * @param Request $request The HTTP request
     * @return int The caller type bitmask
     */
    public function detect(Request $request): int
    {
        $userAgent = $request->header('User-Agent', '');
        $authToken = $request->header('Authorization', '');

        // Check for SEO bot first (highest priority)
        if ($this->isSeoBot($userAgent)) {
            return CallerType::BOT->value;
        }

        // Check for authenticated user
        if ($this->isAuthenticated($authToken)) {
            return CallerType::AUTHENTICATED->value;
        }

        // Default: treat as attacker/unknown caller
        return CallerType::ATTACKER->value;
    }

    /**
     * Check if user agent is a known SEO bot.
     *
     * @param string $userAgent The User-Agent header value
     * @return bool True if the user agent matches a known bot pattern
     */
    private function isSeoBot(string $userAgent): bool
    {
        $userAgentLower = strtolower($userAgent);

        foreach (self::SEO_BOT_PATTERNS as $pattern) {
            if (stripos($userAgentLower, $pattern) !== false) {
                return true;
            }
        }

        return false;
    }

    /**
     * Check if request has valid authentication.
     *
     * @param string $authToken The Authorization header value
     * @return bool True if authentication token is present and valid
     */
    private function isAuthenticated(string $authToken): bool
    {
        if (empty($authToken)) {
            return false;
        }

        // Basic format check: "Bearer <token>" or "Token <token>"
        if (!preg_match('/^(Bearer|Token)\s+\S+$/i', $authToken)) {
            return false;
        }

        // Additional validation could be added here
        // (e.g., token blacklist check, expiry validation)

        return true;
    }

    /**
     * Get human-readable label for caller type.
     *
     * @param int $callerMask The caller type bitmask
     * @return string Human-readable label
     */
    public function getLabel(int $callerMask): string
    {
        return match ($callerMask) {
            CallerType::BOT->value => 'SEO Bot',
            CallerType::AUTHENTICATED->value => 'Authenticated User',
            CallerType::ATTACKER->value => 'Unknown Caller',
            default => 'Unknown Type',
        };
    }
}
